### PR TITLE
Addition of user skill calculation

### PR DESCRIPTION
**Pull request by ramanakumars**
Added a generalized user skill calculation which works in two modes:

* Binary: this is when you don't have multiple classes for the data, but rather just a calculation of whether the user got a classification correct vs incorrect. For example, this is the case for projects like Planet Hunters TESS, where there are no classes to the lightcurve
* k-class: when the user has multiple classes to select from (e.g. through question/survey task) and we need to track the success/failure for individual classes

The mode is determined by passing the `binary` keyword to the reducer. Setting `binary=False` runs the reducer in k-class mode (default).

Success/failure is tracked on FEM using the feedback module. We use a pluck extract to extract the feedback metadata and calculate subject difficulty (#509). This is passed as `relevant_reduction` to the user skill reducer. 

The general framework is to using feedback data to evaluate the skill of the user as they complete gold standard classifications. Correct classifications should increase the user skill, while incorrect gold standard classifications will reduce the user skill. The subject difficulty is used to weight the relative increase/decrease of user skill. Getting difficult subjects (i.e. those that very few users have gotten correctly) will boost user skill much more than getting easy subjects correct. Conversely, getting difficult subjects incorrect should lead to a smaller skill penalty compared to getting easy subjects incorrect.

### Binary problem
The reducer creates a 2x2 confusion matrix (TP, TN, FP, FN), where the TN and FN entries are not defined. TP is the sum of successes weighted by the subject difficulty, while FP is the sum of failures weighted by `1-subject_difficulty`. 

### k-class problem
First, we obtain a full list of labels from both the `true_` keyword from the feedback pluck as well as from the classifications. For `k` labels, we define a `k x k` confusion matrix. The matrix is then filled by looping through the extracts and adding counts to the relevant cell (weighted by difficulty of the subject in that extract). Confusion matrix is constructed using `sklearn.metrics.confusion_matrix` by passing the `subject_difficulty` as the weights for each entry.

For both modes, we set a minimum subject difficulty at `0.05`, so that very easy subjects still have a non-zero effect on the user skill.

### CODE REVIEW

**Comment by ramanakumars** [comment]
cc @CKrawczyk @lcjohnso here's the PR on the user skill calculation for leveling up.

**Review by CKrawczyk — APPROVED** [review]
One small non-blocking comment in the code.  Everything looks reasonable to me (although I have not worked through the math fully).

My one comment is these kinds of user reducers tend to be better off as running reducers because one super-user can have millions of classifications on a project.  As written all of these are passed into the reducer each time, this will result in slow computations and large amounts of data being passed around caesar.

If written as a running reducer only the latest classification is passed in.

File: panoptes_aggregation/reducers/user_skill_reducer.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import numpy as np
+from ..feedback_strategies import FEEDBACK_STRATEGIES
+from sklearn.metrics import confusion_matrix
+import ast
+
+
+@reducer_wrapper(relevant_reduction=True)
+def user_skill_reducer(extracts, relevant_reduction=[], binary=False, null_class='NONE'):
+    if isinstance(binary, str):
+        binary = ast.literal_eval(binary)
```
[the remark this answers is not in the document]
Comment (ramanakumars) [reply panoptes_aggregation/reducers/user_skill_reducer.py]: Yes! Fixed.

**Comment by ramanakumars** [comment]
Yeah we originally wanted the final version to be a running reducer, but that needs a little more understanding of the data flow for this reducer. We were planning to use this PR as a way to run a full simulation of the data flow from existing classifications to see how we can remap it as a running reducer. Any issues with that, or should we work on the running reducer bit for this PR?

**Comment by CKrawczyk** [comment]
That sounds reasonable to me.  I will merge this.
